spec/custom_index_spec.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+# typed: false
+
+describe "Custom index" do
+  let!(:klass) do
+    Class.new(T::Struct) do
+      include Redcord::Base
+
+      attribute :value, T.nilable(Integer)
+      attribute :time_value, T.nilable(Time)
+      attribute :indexed_value, T.nilable(Integer), index: true
+      attribute :other_value, T.nilable(Integer), index: true
+      custom_index :first, [:indexed_value, :time_value, :value]
+
+      if ENV['REDCORD_SPEC_USE_CLUSTER'] == 'true'
+        shard_by_attribute :indexed_value
+      else
+        custom_index :second, [:time_value]
+      end
+
+      def self.name
+        'RedcordSpecModel'
+      end
+    end
+  end
+
+  context 'custom indexes: ' do
+    let!(:time_now) { Time.zone.now}
+    let!(:instance) { klass.create!(indexed_value: 1, time_value: time_now) }
+    let!(:instance_2) { klass.create!(indexed_value: 2, time_value: nil) }
+
+    it 'raises an error when negative value is used for custom index attr' do
+      # shared/lua_helper_methods.erb.lua: adjust_string_length
+      expect { klass.create!(indexed_value: -1) }.to raise_error(Redis::CommandError)
+    end
+
+    it 'raises an error when large numbers (more than 19 digits in decimal notation) used in custom index' do
+      # shared/lua_helper_methods.erb.lua: adjust_string_length
+      expect { klass.create!(indexed_value: 10**19) }.to raise_error(Redis::CommandError)
+      expect { klass.create!(indexed_value: 10**19 - 1) }.to_not raise_error()
+    end
+
+    it 'raises an error when inexisting custom index is queried' do
+      # shared/query_helper_methods.erb.lua: validate_and_parse_query_conditions_custom
+      expect { klass.where(indexed_value: 1).with_index(:third).to_a }.to raise_error(Redis::CommandError)
+    end
+
+    it 'raises error when range query conditions are used not on the last attribute in a query' do
+      # shared/query_helper_methods.erb.lua: validate_and_parse_query_conditions_custom
+      interval = Redcord::RangeInterval.new(min: Time.zone.now)
+      expect {
+        klass.where(indexed_value: 1, time_value: interval, value: 1).with_index(:first).to_a
+      }.to raise_error(Redis::CommandError)
+    end
+
+    it 'raises error when attributes are in incorrect order' do
+      # shared/query_helper_methods.erb.lua: validate_and_parse_query_conditions_custom
+      expect {
+        klass.where(time_value: nil, indexed_value: 1).with_index(:first).to_a
```

Comment:
if you have index like [:a, :b, :c]
query: where(b: val, a: val) could be executed
But I think it becomes harder to explain why you can't use range conditions on :a.
My idea was to communicate to users that with custom indexes order matters and that way I think it would be easier to explain limitations.
Let me know if you strongly believe that commutative property should hold for custom index also.

Reply:
ah I see. Yeah this could be slightly more complex to explain.

Though it could also be confusing since normally the order in the keyword argument lists is not important. So I'm in favor of sorting the args before sending it to redis.

What do you think?

Reply:
I'll try to sort before sending and probably make validation messages for incorrect queries more clear